crates/red_knot_module_resolver/src/path.rs

```diff
@@ -327,28 +432,56 @@ impl<'a> ModuleResolutionPathRefInner<'a> {
     }
 
     #[must_use]
-    fn relativize_path(&self, absolute_path: &'a SystemPath) -> Option<Self> {
-        match self {
-            Self::Extra(root) => absolute_path.strip_prefix(root).ok().and_then(|path| {
-                path.extension()
-                    .map_or(true, |ext| matches!(ext, "py" | "pyi"))
-                    .then_some(Self::Extra(path))
-            }),
-            Self::FirstParty(root) => absolute_path.strip_prefix(root).ok().and_then(|path| {
-                path.extension()
-                    .map_or(true, |ext| matches!(ext, "pyi" | "py"))
-                    .then_some(Self::FirstParty(path))
-            }),
-            Self::StandardLibrary(root) => absolute_path.strip_prefix(root).ok().and_then(|path| {
-                path.extension()
-                    .map_or(true, |ext| ext == "pyi")
-                    .then_some(Self::StandardLibrary(path))
-            }),
-            Self::SitePackages(root) => absolute_path.strip_prefix(root).ok().and_then(|path| {
-                path.extension()
-                    .map_or(true, |ext| matches!(ext, "pyi" | "py"))
-                    .then_some(Self::SitePackages(path))
-            }),
+    fn relativize_path(&self, absolute_path: &StandardLibraryPathRef<'a>) -> Option<Self> {
```

Comment:
It looks like this takes `absolute_path` as a `StandardLibraryPathRef`, not because it really must be a standard library path at all (in many cases we will have to smuggle some other kind of path claiming that it is a `StandardLibraryPathRef`), but simply because that's the only way we can make it possible to ever give a `VendoredPath`?

IMO this makes it super clear that we really do need a refactor of how paths are represented here, this is just a lie in the type signature of the function, and very confusing.

Reply:
Ah, you're right, this is indeed very confusing naming from me. I'll fix this, thanks.

(This is because it was originally envisaged as a general-purpose `FilePathRef` enum in `ruff_db`, but Micha asked me to move it to this crate for now. I renamed it without thinking too much about it when I moved it into this crate, but `FilePathRef` probably is better after all.)

Separately, I do still also agree that your suggested refactoring from the other PR sounds good, and still have it on my to-do list to look into it.